service/matching/taskReader.go

```diff
@@ -421,3 +376,90 @@ func (tr *taskReader) newDispatchContext(isolationGroup string) (context.Context
 }
+
+func (tr *taskReader) dispatchSingleTaskFromBuffer(isolationGroup string, taskInfo *persistence.TaskInfo) (breakDispatchLoop bool) {
+	// retry loop for dispatching a single task
+	for {
+		task := newInternalTask(taskInfo, tr.completeTask, types.TaskSourceDbBacklog, "", false, nil, isolationGroup)
+		dispatchCtx, cancel := tr.newDispatchContext(isolationGroup)
+		timerScope := tr.scope.StartTimer(metrics.AsyncMatchLatencyPerTaskList)
+		err := tr.dispatchTask(dispatchCtx, task)
+		timerScope.Stop()
+		cancel()
+		if err == nil {
+			return
+		}
+		if err == context.Canceled {
+			tr.logger.Info("Tasklist manager context is cancelled, shutting down")
+			breakDispatchLoop = true
+			return
+		}
+		if err == context.DeadlineExceeded {
+			// it only happens when isolation is enabled and there is no pollers from the given isolation group
+			// if this happens, we don't want to block the task dispatching, because there might be pollers from
+			// other isolation groups, we just simply continue and dispatch the task to a new isolation group which
+			// has pollers
+			tr.logger.Warn("Async task dispatch timed out", tag.IsolationGroup(isolationGroup))
+			tr.scope.IncCounter(metrics.AsyncMatchDispatchTimeoutCounterPerTaskList)
+
+			// the idea here is that by re-fetching the isolation-groups, if something has shifted
+			// it will get a new isolation group to be placed. If it needs re-routing, then
+			// this will be the new routing destination.
+			group, err := tr.getIsolationGroupForTask(tr.cancelCtx, taskInfo)
+			if err != nil {
+				// it only errors when the tasklist is a sticky tasklist and
+				// the sticky pollers are not available, in this case, we just complete the task
+				// and let the decision get timed out and rescheduled to non-sticky tasklist
+				if err == _stickyPollerUnavailableError {
+					tr.completeTask(taskInfo, nil)
+				} else {
+					// it should never happen, unless there is a bug in 'getIsolationGroupForTask' method
+					tr.logger.Error("taskReader: unexpected error getting isolation group", tag.Error(err))
+					tr.completeTask(taskInfo, err)
+				}
+				break
+			}
+			if group == isolationGroup {
+				continue
+			} else {
+				// if there is no poller in the isolation group or the isolation group is drained,
+				// we want to redistribute the tasks to other isolation groups in this case to drain
+				// the backlog
+				select {
+				case <-tr.cancelCtx.Done():
+					// the task reader is shutting down
+					breakDispatchLoop = true
+					return
+				case tr.taskBuffers[group] <- taskInfo:
+					// successful redirect
+					tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectCounter)
+					tr.logger.Warn("some tasks were redirected to another isolation group", tag.IsolationGroup(group),
+						tag.WorkflowRunID(taskInfo.RunID),
+						tag.WorkflowID(taskInfo.WorkflowID),
+						tag.TaskID(taskInfo.TaskID),
+						tag.WorkflowDomainID(taskInfo.DomainID),
+					)
+				default:
```

Comment:
This, here, is the single line hotfix basically. The default should prevent any attempts to put stuff on the channel from deadlocking if it's full.